### ISSUE TITLE
Looping events

### ISSUE THREAD

**Issue by ErikBjare** [Homebrain/Homebrain#9]
To make my case clear I'll start with a few scenarios.
### Scenario 1

> There is one agent A which subscribes to "test" events. When it has received this event it creates a new event "test".

This will cause an infinite message loop, (A → A → A → ⋯) which we can fix easily enough. This fix will be presented in the next scenario.
### Scenario 2

> There is one agent A which subscribes to "test" events. When it has received this event it creates a new event "test" with the data field `{"visited": True}` which prevents it from firing a new event.

This will work wonderfully, no runaway loops, but if we introduce a new agent...

> Agent B also subscribes to all "test" events, but discards the "visited" field or passes in another data field entirely. 

When this message is sent back to the dispatcher it will again be caught up by agent A, which don't find the data.visited field and creates another event, closing the loop (A → B → A → B → ⋯).
## Discussion

This might not be a huge problem, since it should be rare and usually the fault of the person implementing the first or the second agent. But it is likely to appear sooner rather than later, this is not the first time I've brought this up, I noted it as a major issue with giving events the ability to loop back through the dispatcher instead of building the message passing as a directed acyclic graph. I still think the current design has a few advantages, but it might be reasonable to enforce a few rules when it comes to how things can be chained.
## Proposed solution

The only two I can come up with are:
- Enforce all events to have their ancestors available for query, and query them when a loop is suspected.
- Change to an architecture which does message passing through a DAG structure.

These two are both pretty bothersome, but the second eliminates the problem entirely and cleanly. What are the disadvantages to enforce such a structure in the dispatchers chain logic?

**Comment by ErikBjare** [comment]
@vlthr I like that solution, I don't think this issue deserved the attention I gave it. The `event["history"]` is appealing to me, at least as a debugging mechanism.

**Comment by vlthr** [comment]
I agree - history would be interesting for debugging/logging even if it is not used for actual loop detection
